### PR TITLE
Add GotFilled

### PR DESCRIPTION
**Pull request by Opacity000**
- Content is dynamically loaded, so it needs to use CDP
- There is an "age > 18" prompt on the website, it's bypassed using the `gotfilled_warning_ok` cookie

### CODE REVIEW

File: scrapers/GotFilled.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+xPathScrapers:
+  sceneScraper:
+    scene:
+      Title: //*[@id="info"]/h2/text()
+      Date:
+        selector: //*[@id="info"]/div[4]/div[3]/h5/text()
+        postProcess:
+          - parseDate: Jan 2, 2006
+      Details: //*[@id="info"]/div[1]/text()
```
Comment (Opacity000) [inline scrapers/GotFilled.yml:22]: Can be refactored to `//*[@id="info"]/div[@class="desc"]/text()`

File: scrapers/GotFilled.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      Performers:
+        Name: //*[@id="info"]/div[2]/h3/a/text()
```
Comment (Opacity000) [inline scrapers/GotFilled.yml:24]: Can be refactored to `//*[@id="info"]/div[@class="video-model"]/h3/a/text()`

File: scrapers/GotFilled.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      Tags:
+        Name: //*[@id="info"]/div[5]/div[2]/p/a/text()
+      Image:
+      # Images are a div with an inline style containing a background-image with
+      # a url
+      # Inline style exemple: `background-image: url("{IMAGE_URL}");`
+        selector: //*[@id="vjs_video_3"]/div[3]/@style
```
Comment (Opacity000) [inline scrapers/GotFilled.yml:31]: Can be refactored to `//*[@id="vjs_video_3"]/div[@class="vjs-poster"]/@style`

File: scrapers/GotFilled.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          - replace:
+            - regex: .*url\("|"\);
+              with: ""
+      Studio:
+        Name:
+          fixed: GotFilled
+
+  performerScraper:
+    common:
+      $details: //*[@id="root"]/main/div/div[2]/div[2]
```
Comment (Opacity000) [inline scrapers/GotFilled.yml:42]: Can be refactored to `//*[@id="root"]/main/div/div[contains(@class, "model-info")]/div[contains(@class, "details")]`

File: scrapers/GotFilled.yml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      Weight:
+        selector: $details/div/ul/li[7]/p/text()
+        postProcess:
+          - replace:
+            - regex: " lbs"
+              with: ""
+          - lbToKg: true
+      Measurements: $details/div/ul/li[4]/p/text()
+      Details: $details/p
+      Image: //*[@id="root"]/main/div/div[2]/div[1]/img/@src
```
Comment (Opacity000) [inline scrapers/GotFilled.yml:61]: Can be refactored to `//*[@id="root"]/main/div/div[contains(@class, "model-info")]/div[contains(@class, "thumbnail")]/img/@src`

**Review by Opacity000 — COMMENTED** [review]
What do you think of these changes ?

**Comment by Belleyy** [comment]
I saw that `bjraw` look like using the same layout, so maybe adding the URL in the scraper can maybe work.

File: scrapers/GotFilled.yml [hunk 6]
```diff
@@ -64,10 +64,4 @@ xPathScrapers:
 
 driver:
   useCDP: true
-  cookies:
-    - Cookies:
-      - Name: gotfilled_warning_ok
-        Domain: gotfilled.com
-        Value: "1"
-        Path: /
```
Comment (Opacity000) [inline scrapers/GotFilled.yml]: The "age > 18" prompt does not prevent the scene and performer pages from loading, mb

**Review by Opacity000 — COMMENTED** [review]
I changed the name to `KB Productions` because that's the company that owns the website (and probably the studio?).

I've checked the other websites they own, and there are many, some even have their own scrapers already (e.g. Nympho.yml, Hitzefrei.yml).

Should I revert this commit and keep `GotFilled` as the name file and name value ?

**Comment by Belleyy** [comment]
> Should I revert commit 277ef66 and keep GotFilled as the name file and name value ?

I personally ok with `KB Productions`, if they create more studio they can maybe use the same layout.

File: scrapers/GotFilled.yml (hunk 1 above)
Comment (bnkai) [reply scrapers/GotFilled.yml]: yeah also safer

File: scrapers/GotFilled.yml (hunk 2 above)
Comment (bnkai) [reply scrapers/GotFilled.yml]: same here the refactored is bettter/safer

File: scrapers/GotFilled.yml (hunk 3 above)
Comment (bnkai) [reply scrapers/GotFilled.yml]: same here the refactored is bettter/safer, you can drop the comments also

File: scrapers/GotFilled.yml (hunk 5 above)
Comment (bnkai) [reply scrapers/GotFilled.yml]: yeah safer/better

File: scrapers/KBProductions.yml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      - gotfilled.com/tour/models/
+      - bjraw.com/tour/models/
+    scraper: performerScraper
+
+xPathScrapers:
+  sceneScraper:
+    scene:
+      Title: //*[@id="info"]/h2/text()
+      Date:
+        selector: //*[@id="info"]/div[4]/div[3]/h5/text()
```
Comment (bnkai) [inline scrapers/KBProductions.yml:21]: a safer selector can be used here
```yaml
        selector: //*[@id="info"]//div[starts-with(h4,"Publish")]/h5
```

**Comment by Opacity000** [comment]
I guess we're all done

File: scrapers/KBProductions.yml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      Name: $details/h2/text()
+      Birthdate: $details/div/ul/li[1]/p/text()
+      HairColor: $details/div/ul/li[6]/p/text()
+      EyeColor: $details/div/ul/li[5]/p/text()
+      Height:
+        selector: $details/div/ul/li[8]/p/text()
+        postProcess:
+          - feetToCm: true
+      Weight:
+        selector: $details/div/ul/li[7]/p/text()
```
Comment (bnkai) [inline scrapers/KBProductions.yml:58]: some more ?
```yaml
   Birthdate: $details/div/ul/li[h3[text()="Birthdate"]]/p/text()
      HairColor: $details/div/ul/li[h3[text()="Hair"]]/p/text()
      EyeColor: $details/div/ul/li[h3[text()="Eyes"]]/p/text()
      Height:
        selector: $details/div/ul/li[h3[text()="Height"]]/p/text()
        postProcess:
          - feetToCm: true
      Weight:
        selector: $details/div/ul/li[h3[text()="Weight"]]/p/text()
```

File: scrapers/KBProductions.yml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        postProcess:
+          - replace:
+            - regex: " lbs"
+              with: ""
+          - lbToKg: true
+      Measurements: $details/div/ul/li[4]/p/text()
```
Comment (bnkai) [inline scrapers/KBProductions.yml:64]: and here ?
```yaml
      Measurements: $details/div/ul/li[h3[text()="Measurements"]]/p/text()
```

**Comment by bnkai** [comment]
Also i needed to add back the cookies as i was getting some timeouts ....
```yaml
driver:
  useCDP: true
  cookies:
    - Cookies:
      - Name: gotfilled_warning_ok
        Domain: gotfilled.com
        Value: "1"
        Path: /
    - Cookies:
      - Name: bj-raw_warning_ok
        Domain: bjraw.com
        Value: "1"
        Path: /
```
The sites btw are slow and using cdp doesnt help....

File: scrapers/KBProductions.yml (hunk 8 above)
Comment (Opacity000) [reply scrapers/KBProductions.yml:58]: Oh, right. I've missed them 😅

**Comment by Opacity000** [comment]
> Also i needed to add back the cookies as i was getting some timeouts ....

There was no problem on my side, weird. I deleted them to avoid having to add a cookie for each site, but I'm gonna add them back.

> The sites btw are slow and using cdp doesnt help....

I know right ? And the popup asking for confirmation to enter the site doesn't help either.

**Comment by bnkai** [comment]
After trying to verify the latest changes and getting timeouts allover i switched to a python scraper to get the data from the script in the page. This works faster and more reliably

The ending scraper looks like this

KBProductions.py
```py
import json
import os
import re
import sys

try:
    import py_common.log as log
except ModuleNotFoundError:
    print(
        "You need to download the folder 'py_common' from the community repo (CommunityScrapers/tree/master/scrapers/py_common)",
        file=sys.stderr)
    sys.exit(1)

import requests
from bs4 import BeautifulSoup


def get_from_url(url_to_parse):
    m = re.match(
        r'https?://(?:www\.)?((\w+)\.com)/tour/(?:videos|upcoming)/(\d+)/([a-z0-9-]+)',
        url_to_parse)
    if m is None:
        return None, None, None, None
    return m.groups()


def make_request(request_url, origin_site):
    requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += 'HIGH:!DH:!aNULL'
    requests.packages.urllib3.disable_warnings()

    try:
        requests.packages.urllib3.contrib.pyopenssl.DEFAULT_SSL_CIPHER_LIST += 'HIGH:!DH:!aNULL'
    except AttributeError:
        # no pyopenssl support used / needed / available
        pass
    try:
        r = requests.get(
            request_url,
            headers={
                'User-Agent':
                'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:79.0) Gecko/20100101 Firefox/79.0',
                'Origin': origin_site,
                'Referer': request_url
            },
            timeout=(3, 6),
            verify=False)
    except requests.exceptions.RequestException as e:
        return None, e

    if r.status_code == 200:
        return r.text, None
    return None, f"HTTP Error: {r.status_code}"


def fetch_page_json(page_html):
    matches = re.findall(r'window\.__DATA__ = (.+)$', page_html, re.MULTILINE)
    return json.loads(matches[0]) if matches else None


def main():
    stdin = sys.stdin.read()
    log.debug(stdin)
    fragment = json.loads(stdin)

    if not fragment['url']:
        log.error('No URL entered.')
        sys.exit(1)
    url = fragment['url'].strip()
    site, studio, sid, slug = get_from_url(url)
    if site is None:
        log.error('The URL could not be parsed')
        sys.exit(1)
    response, err = make_request(url, f"https://{site}")
    if err is not None:
        log.error('Could not fetch page HTML', err)
        sys.exit(1)
    j = fetch_page_json(response)
    #log.debug(response)
    #log.debug(j)
    if j is None or j.get("video") is None:
        log.error('Could not find JSON on page')
        sys.exit(1)

    scene = j["video"]
    #log.debug(scene)

    scrape = {}
    scrape['studio'] = {'name': studio}
    if scene.get('title'):
        scrape['title'] = scene['title']
    if scene.get('release_date'):
        scrape['date'] = scene['release_date'][:10]
    if scene.get('description'):
        details = BeautifulSoup(scene['description'], "html.parser").get_text()
        scrape['details'] = details
    if scene.get('models'):
        models = []
        for m in scene['models']:
            if m.get('name'):
                models.append(m['name'])
        scrape['performers'] = [{'name': x} for x in models]
    if scene.get('tags'):
        tags = []
        for t in scene['tags']:
            if t.get('name'):
                tags.append(t['name'])
        scrape['tags'] = [{'name': x} for x in tags]
    if scene.get('extra_thumbs'):
        # available image endings
        # ================
        #_player.jpg
        #_playermobile.jpg
        #_portrait1.jpg
        #_portrait2.jpg
        #_scene.jpg
        #_scenemobile.jpg
        img = None
        for i in scene['extra_thumbs']:
            if i.endswith("_player.jpg"):
                image = i
                break
        if img is None:
            img = scene['extra_thumbs'][0]
        scrape['image'] = img
    print(json.dumps(scrape))


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        log.error(e)

```

KBProductions.yml
```yaml
name: "KB Productions"
sceneByURL:
  - url:
      - bjraw.com/tour/upcoming/
      - bjraw.com/tour/videos/
      - gotfilled.com/tour/upcoming/
      - gotfilled.com/tour/videos/

    action: script
    script:
      - python3
      - KBProductions.py
# Last Updated January 16, 2022
```

**Comment by Opacity000** [comment]
> After trying to verify the latest changes and getting timeouts allover i switched to a python scraper to get the data from the script in the page. This works faster and more reliably

It is then possible to implement searches if we use python

**Comment by bnkai** [comment]
> > After trying to verify the latest changes and getting timeouts allover i switched to a python scraper to get the data from the script in the page. This works faster and more reliably
> 
> It is then possible to implement searches if we use python

I think it was possible even without python, given that they have a search endpoint
I can have a look when i look at the performer scraper part
Can you verify that the scene scraper works ok?

**Comment by Opacity000** [comment]
> I think it was possible even without python, given that they have a search endpoint

Is it possible to set multiple URLs for searches without using python ?

> Can you verify that the scene scraper works ok?

It works as expected, everything's ok

**Comment by bnkai** [comment]
> Is it possible to set multiple URLs for searches without using python ?
Even with python if you want 2 different site  searches you would need 2 separate yaml files so that you can have different names to select from the scaper dropdown. It's just that with script scrapers you can keep all the actual scraping in a single py file

**Comment by Opacity000** [comment]
Yeah, so python avoids code duplication, and so it avoids having to rewrite multiple files if something needs to be rewritten.
That's what I meant when I said "It is then possible to implement searches if we use python", I just expressed myself very badly.

**Comment by bnkai** [comment]
Updated files with performerByUrl support

KBProductions.yml
```yaml
name: "KB Productions"
sceneByURL:
  - url:
      - bjraw.com/tour/upcoming/
      - bjraw.com/tour/videos/
      - gotfilled.com/tour/upcoming/
      - gotfilled.com/tour/videos/

    action: script
    script:
      - python3
      - KBProductions.py
      - scene
performerByURL:
  - url:
      - bjraw.com/tour/models/
      - gotfilled.com/tour/models/
    action: script
    script:
      - python3
      - KBProductions.py
      - performer
# Last Updated January 16, 2022
```
KBProductions.py
```py
import json
import os
import re
import sys

try:
    import py_common.log as log
except ModuleNotFoundError:
    print(
        "You need to download the folder 'py_common' from the community repo (CommunityScrapers/tree/master/scrapers/py_common)",
        file=sys.stderr)
    sys.exit(1)

import requests
from bs4 import BeautifulSoup


def get_from_url(url_to_parse):
    m = re.match(
        r'https?://(?:www\.)?((\w+)\.com)/tour/(?:videos|upcoming|models)/(\d+)/([a-z0-9-]+)',
        url_to_parse)
    if m is None:
        return None, None, None, None
    return m.groups()


def make_request(request_url, origin_site):
    requests.packages.urllib3.util.ssl_.DEFAULT_CIPHERS += 'HIGH:!DH:!aNULL'
    requests.packages.urllib3.disable_warnings()

    try:
        requests.packages.urllib3.contrib.pyopenssl.DEFAULT_SSL_CIPHER_LIST += 'HIGH:!DH:!aNULL'
    except AttributeError:
        # no pyopenssl support used / needed / available
        pass
    try:
        r = requests.get(
            request_url,
            headers={
                'User-Agent':
                'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:79.0) Gecko/20100101 Firefox/79.0',
                'Origin': origin_site,
                'Referer': request_url
            },
            timeout=(3, 6),
            verify=False)
    except requests.exceptions.RequestException as e:
        return None, e

    if r.status_code == 200:
        return r.text, None
    return None, f"HTTP Error: {r.status_code}"


def fetch_page_json(page_html):
    matches = re.findall(r'window\.__DATA__ = (.+)$', page_html, re.MULTILINE)
    return json.loads(matches[0]) if matches else None


def scrape_scene(page_json, studio):
    if page_json.get("video") is None:
        log.error('Could not find  scene in JSON data')
        sys.exit(1)

    scene = page_json["video"]

    scrape = {}
    scrape['studio'] = {'name': studio}
    if scene.get('title'):
        scrape['title'] = scene['title']
    if scene.get('release_date'):
        scrape['date'] = scene['release_date'][:10]
    if scene.get('description'):
        details = BeautifulSoup(scene['description'], "html.parser").get_text()
        scrape['details'] = details
    if scene.get('models'):
        models = []
        for m in scene['models']:
            if m.get('name'):
                models.append(m['name'])
        scrape['performers'] = [{'name': x} for x in models]
    if scene.get('tags'):
        tags = []
        for t in scene['tags']:
            if t.get('name'):
                tags.append(t['name'])
        scrape['tags'] = [{'name': x} for x in tags]
    if scene.get('extra_thumbs'):
        # available image endings
        # ================
        #_player.jpg
        #_playermobile.jpg
        #_portrait1.jpg
        #_portrait2.jpg
        #_scene.jpg
        #_scenemobile.jpg
        img = None
        for i in scene['extra_thumbs']:
            if i.endswith("_player.jpg"):
                image = i
                break
        if img is None:
            img = scene['extra_thumbs'][0]
        scrape['image'] = img
    print(json.dumps(scrape))


def get_dict_value(d: dict, v: str):
    if d.get(v):
        return d[v]
    return None


def scrape_performer(page_json):
    if page_json.get("model") is None:
        log.error('Could not find performer in JSON data')
        sys.exit(1)

    performer = page_json["model"]
    scrape = {}

    scrape['name'] = get_dict_value(performer, 'name')
    scrape['gender'] = get_dict_value(performer, 'gender')
    scrape['image'] = get_dict_value(performer, 'thumb')

    if performer.get('attributes'):
        pa = performer['attributes']
        if pa.get('bio'):
            scrape['details'] = get_dict_value(pa['bio'], 'value')
        if pa.get('birthdate'):
            scrape['birthdate'] = get_dict_value(pa['birthdate'], 'value')
        if pa.get('measurements'):
            scrape['measurements'] = get_dict_value(pa['measurements'],
                                                    'value')
        if pa.get('eyes'):
            scrape['eye_color'] = get_dict_value(pa['eyes'], 'value')
        if pa.get('height'):
            height_ft = get_dict_value(pa['height'], 'value')
            if height_ft:
                h = re.match(r'(\d+)\D(\d+).+', height_ft)
                if h:
                    h_int = int(
                        round((float(h.group(1)) * 12 + float(h.group(2))) *
                              2.54))  # ft'inches to cm
                    scrape['height'] = f"{h_int}"
        if pa.get('weight'):
            weight_lb = get_dict_value(pa['weight'], 'value')
            if weight_lb:
                w = re.match(r'(\d+)\slbs', weight_lb)
                if w:
                    w_int = int(round(float(w.group(1)) / 2.2046))  # lbs to kg
                    scrape['weight'] = f"{w_int}"
        if pa.get('hair'):
            scrape['hair_color'] = get_dict_value(pa['hair'], 'value')
    print(json.dumps(scrape))


def main():
    stdin = sys.stdin.read()
    log.debug(stdin)
    fragment = json.loads(stdin)

    if not fragment['url']:
        log.error('No URL entered.')
        sys.exit(1)
    url = fragment['url'].strip()

    site, studio, el_id, slug = get_from_url(url)
    if site is None:
        log.error('The URL could not be parsed')
        sys.exit(1)

    response, err = make_request(url, f"https://{site}")
    if err is not None:
        log.error(f"Could not fetch page HTML: {err}")
        sys.exit(1)

    j = fetch_page_json(response)
    if j is None:
        log.error('Could not find JSON on page')
        sys.exit(1)

    if len(sys.argv) == 0 or sys.argv[1] == "scene":
        scrape_scene(page_json=j, studio=studio)
    elif sys.argv[1] == "performer":
        scrape_performer(j)


if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        log.error(e)
```

**Comment by bnkai** [comment]
Merging this, will update with the search functionality at a later date

**Pull request merged**